public/friendlycode/js/fc/publisher.js

```diff
@@ -0,0 +1,203 @@
+define(function(require) {
+  var $ = require("jquery");
+  var host;
+  var TEXT_PUBLISH = "Publish";
+  var TEXT_PUBLISHING = "Publishing...";
+  var TEXT_UNPUBLISH = "Unpublish";
+  var TEXT_UNPUBLISHING = "Unpublishing...";
+  var TEXT_UPDATE_PUBLISH = "Update published version";
+
+  function togglePublicVisibility() {
+    var toggle = this.interface.projectVisibility;
+    var ellipse = $({ cx: toggle.attr("cx") });
+    var cx = 31;
+    var fill = "#06a050";
+    var animationOptions = {
+      duration: 250,
+      step: function(now) {
+        toggle.attr("cx", now);
+      }
+    };
+
+    this.public = !this.public;
+
+    if(!this.public) {
+      cx = 11;
+      fill = "#7C7C7C";
+    }
+
+    ellipse.animate({ cx: cx }, animationOptions);
+    toggle.css("fill", fill);
+  }
+
+  function unpublishedChangesPrompt() {
+    var interface = this.interface;
+    var publish = this.handlers.publish;
+    interface.published.changed.removeClass("hide");
+    interface.buttons.update
+      .off("click", publish)
+      .on("click", publish);
+  }
+
+  function Publisher(editor, options) {
+    this.bramble = editor;
+    this.fsync = options.sync;
+    host = options.appUrl;
+    this.csrfToken = $("meta[name='csrf-token']").attr("content");
+    this.interface = {
+      buttons: {
+        publish: $("#publish-button-publish"),
+        update: $("#publish-button-update"),
+        unpublish: $("#publish-button-unpublish")
+      },
+      description: $("#publish-details > textarea.publish-description"),
+      projectVisibility: $("#publish-public-gallery-toggle > ellipse"),
+      published: {
+        link: $("#publish-link > a"),
+        changed: $("#publish-changes"),
+        container: $("#publish-live")
+      }
+    };
+  }
+
+  Publisher.prototype.init = function(project) {
+    var publisher = this;
+    var interface = publisher.interface;
+    publisher.public = true;
```

Comment:
This seems to manage the project's state as well as the state of publishing itself. Are you sure we want to mix those? It makes this somewhat confusing.